superset/models/core.py

```diff
@@ -840,61 +863,130 @@ def inspector(self):
         engine = self.get_sqla_engine()
         return sqla.inspect(engine)
 
-    def all_table_names(self, schema=None, force=False):
-        if not schema:
-            if not self.allow_multi_schema_metadata_fetch:
-                return []
-            tables_dict = self.db_engine_spec.fetch_result_sets(
-                self, 'table', force=force)
-            return tables_dict.get('', [])
-
-        extra = self.get_extra()
-        medatada_cache_timeout = extra.get('metadata_cache_timeout', {})
-        table_cache_timeout = medatada_cache_timeout.get('table_cache_timeout')
-        enable_cache = 'table_cache_timeout' in medatada_cache_timeout
-        return sorted(self.db_engine_spec.get_table_names(
-            inspector=self.inspector,
-            db_id=self.id,
-            schema=schema,
-            enable_cache=enable_cache,
-            cache_timeout=table_cache_timeout,
-            force=force))
-
-    def all_view_names(self, schema=None, force=False):
-        if not schema:
-            if not self.allow_multi_schema_metadata_fetch:
-                return []
-            views_dict = self.db_engine_spec.fetch_result_sets(
-                self, 'view', force=force)
-            return views_dict.get('', [])
+    @cache_util.memoized_func(
+        key=lambda *args, **kwargs: 'db:{db_id}:schema:None:table_list'.format(
+            db_id=kwargs.get('db_id')),
+        use_tables_cache=True)
+    def all_table_names_in_database(self, db_id=None, cache=False,
+                                    cache_timeout=None, force=False):
+        """
+        Parameters need to be passed as keyword arguments.
+        If cache=True, db_id must be passed in for setting cache key
```

Comment:
Oh, you want the database id to be part of the key. In that case, I think it's better to create a new decorator reusing `cache_util.memoized_func`, that extracts the `id` from `self`. Here's a simple example showing the technique:

```python
def instance_memo(attribute, cache={}):
    def decorator(func):
        def decorated(self, *args, **kwargs):
            # in this example, key is just `Database.id`
            key = getattr(self, attribute)
            if key not in cache:
                print('Writing to cache')
                cache[key] = func(self, *args, **kwargs)
            else:
                print('Reading from cache')
            return cache[key]
        return decorated
    return decorator


class Database:
    def __init__(self, id):
        self.id = id

    @instance_memo('id')
    def method(self):
        pass
```

Calling it we get:

```python
>>> db1 = Database(1)
>>> db1.method()
Writing to cache
>>> db1.method()
Reading from cache
>>> db1.method()
Reading from cache
>>> db2 = Database(2)
>>> db2.method()
Writing to cache
>>> db2.method()
Reading from cache
>>> db2.method()
Reading from cache
```